File: iOS/Sources/Preview/Sources/BeaglePreview.swift
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import UIKit

import BeagleUI

public class BeaglePreview {

    // MARK: Static

    public static func present(in viewController: UIViewController) {
        self.shared.present(in: viewController)
    }

    deinit {
        self.viewController = nil
        self.connection.disconnect()
    }

    // MARK: Private

    private static let shared = BeaglePreview()

    private var connection = ConnectionHandler()
    private var viewController: BeaglePreviewViewController?

    private init() {
        self.connection.delegate = self
    }

    private func present(in presentingViewController: UIViewController) {

        let viewController = BeaglePreviewViewController()
        viewController.modalPresentationStyle = .fullScreen
        presentingViewController.present(viewController, animated: true, completion: { [weak self] in
            guard let self = self else { return }
            self.connection.connect()
        })

        self.viewController = viewController
    }

}

extension BeaglePreview: ConnectionHandlerDelegate {

    func onLayoutChange(_ json: String) {
        viewController?.reloadScreen(with: json)
    }
}

File: android/android-preview/src/main/java/br/com/zup/beagle/preview/BeaglePreview.kt
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package br.com.zup.beagle.preview

import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.WebSocket
import okhttp3.WebSocketListener

internal const val PREVIEW_ENDPOINT = "http://10.0.2.2"
internal const val DEFAULT_PORT = "9721"
internal const val CLOSE_CODE = 1001
internal const val CLOSE_REASON = "onDestroy"

class BeaglePreview(host: String? = null, private val okHttpClient: OkHttpClient = OkHttpClient()) {

    private val request: Request = Request
        .Builder()
        .url("${host ?: PREVIEW_ENDPOINT}:$DEFAULT_PORT")
        .build()
    private lateinit var webSocket: WebSocket

    fun start(webSocketListener: WebSocketListener) {
        webSocket = okHttpClient.newWebSocket(request, webSocketListener)
    }

    fun close() {
        webSocket.close(code = CLOSE_CODE, reason = CLOSE_REASON)
    }
}
